ls1intum/Artemis
# Feature/programming exercise/lock student repositories after due date

### Checklist

### Motivation and Context



This addresses https://github.com/ls1intum/Artemis/issues/847.
If the buildAndTestAfterDueDate is set, we schedule a task when the due date passes to lock the student repositories so they can't change their solution anymore.

### Description


This has a couple of layers, all only active if the buildAfterDueDate is set:
- Integrate scheduling task to lock the student repositories when the due date passes
- Disable the code-editor UI when the due date passes
- Show an info element in the code-editor so the student understands the repository is locked (when the due date has passed)
- ~Inverse Result.rated determination: When the buildAfterDueDate is set, this means that results before the due date are unrated and after it are rated.~ => I decided to move this to another PR: https://github.com/ls1intum/Artemis/pull/869

:information_source: I tested this with ~1000 student participations on the test server and didn't notice any issues.

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration > Programming Exercise > Create
3. Set Due Date and BuildAfterDueDate in future, save
4. Make a student submission (start exercise, submit) => should work, returns a result
5. Wait for the dueDate to pass
6. Go to the code-editor: You should not be able to submit anymore, a warning should be displayed
7. Go to bitbucket: The student should no longer have write permissions.

### Screenshots



Note 1: There is a 20 second delay in the middle of the GIF until the scheduler runs (sorry!)
Note 2: As mentioned above, I decided to move the changes to the graded / ungraded logic to another PR, so the graded state will be inverted when you try it out.
Note 3: A student has no access in Bitbucket to the project settings (so they can't obviously give themselves writing permissions after they were removed :D)

![buildAndTestAfterDueDateLock](https://user-images.githubusercontent.com/28230611/65817552-d0ab1700-e208-11e9-88f4-3b40a789a596.gif)
